A mechanism to associate a tactic with an implicit parameter via an attribute

nikswamy opened the pull request:
This is still quite an experimental feature, extending the syntax of implicit arguments like so:

```
let bind (a:Type) (b:Type)
  (#[@resolve_framing] pre_f:pre_t)
  (#[@resolve_framing] post_f:post_t a)
  (#[@resolve_framing] post_g:post_t b)
  (f:repr a pre_f post_f)
  (g:(x:a -> repr b (post_f x) post_g))
: repr b pre_f post_g
```

which dispatches the resolution of those implicit arguments to a tactic in scope with an attribute like so:

```
[@(resolve_implicits)
  (resolve_framing)]
let resolve () : Tac unit =
```

This is definitely still very experimental and not recommended for general use. However, it relies on changes from the syntax on up and touches many parts of the typechecker. As such, I would like to get it to master to ease the maintenance burden and to enable using it in many other branches that fork from master as we work on Steel.

I don't think this should disturb any existing functionality, but it would be good to have another set of eyes, if possible.

aseemr commented:
Thanks @nikswamy! This would be very useful for me to do some testing on the `_aseem_steel_framing` branch where I am experimenting with the 2 effects story.

I will review this in detail soon, but have some high-level comments for now:

- Should the syntax of these binder attributes match with the new `@@` syntax?
- I think the code could use some comments (e.g. new fields in `Syntax`, `Rel`, `Env`, code in Rel that does the dispatching after bucketizing, etc.). I can make sense of what they are doing but some comments would also be nice.
- Should we remove the files `ulib/experimental/SteelT.*` from the PR?
- Can we also cleanup the previously added `resolve_all_implicits_hook` in this PR itself? The `Rel` support for this never made it to the master branch, but there are remnants of it in `Env`, `Tactics`, `Pervasives` etc. that we should get rid of.

R1kM commented:
I agree with Aseem's high-level comments.
Regarding Steel files in particular, I think all the FramingBind, Repro and so on should not be included in master. They are remnants of experiments that didn't work, and should hopefully be replaced by the new 2-effect system.
I'm also not completely sure whether DeferredImplicits.fs(i) should be included, as it is extremely Steel-specific WIP for now. Maybe we could keep the machinery and interface as it is, but replace the implementation by a no-op for now? On the fence about this one.
Thanks!

mtzguido commented on src/syntax/FStar.Syntax.Resugar.fs:
```diff
@@ -1095,16 +1095,19 @@ and resugar_arg_qual env (q:option<S.arg_qualifier>) : option<(option<A.arg_qual
   | Some (S.Implicit true) -> A.Nothing // We don't have syntax for inaccessible arguments
-  | Some (S.Meta t) -> A.HashBrace (resugar_term' env t)
+  | Some (S.Meta (S.Arg_qualifier_meta_tac t)) -> A.HashBrace (resugar_term' env t)
+  | Some (S.Meta (S.Arg_qualifier_meta_attr t)) -> A.Nothing //TODO: Review this 03/31
```
I think it's indeed fine to return `A.Nothing`, since these implicits come only from applications. Then the case above also seems useless.. I think I added it, I could remove it later.

mtzguido commented on src/typechecker/FStar.TypeChecker.Rel.fs:
```diff
@@ -3860,22 +3933,137 @@ let resolve_implicits' env must_total forcelax g =
+                    guard = %s\n"
+                    (guard_to_string env g);
+
+    let solve_goals_with_tac (imps:implicits) (tac:sigelt) =
```
Would be nice to split this whole chunk into a separate function, so `force_trivial_guard` remains more readable, but we can do that in a later cleanup.

nikswamy replied:
That's a good suggestion. I can do that before merging.

nikswamy commented:
Thanks for taking a look promptly! Much appreciated.

Some responses:

* Regarding the syntax: we currently support only adding a single attribute to a binder. That makes it different than `[@@ e1;e2]` etc. Maybe we should allow multiple attributes here as well. i.e., one would write

`(#[@@ resolve_implicits foobar; some; other; attrributes] x) ... `

which would dispatch solving this implicit to a tactic with the attribute `resolve_implicits foobar`. And the rest of the attributes would be uninterpreted for now, but we may have uses for them in the future.

* Comments, cleanup, removing the old hook etc: Yep, will do.

* Removing DeferredImplicits: Fine with that too. Maintaining and restoring that separately on a branch isn't much work since there's no potentially conflicting churn there. But, that essentially would also mean that the SteelT files go ... since without this sorting, those files will no longer typecheck.

But without those SteelT files, we don't really have existing test cases for this new feature. Makes me wonder if this is worth getting to master at all, until we have something that really uses it successfully.

@aseemr, what do you think about continuing your work by merging this into your branch, rather than using master as a synchronization point?

aseemr commented on src/parser/parse.fsy:
```diff
@@ -1024,7 +1027,10 @@ aqual:
   HASH LBRACK thunk_tmNoEq_ RBRACK
     {let (_1, _2, t, _4) = ((), (), $3, ()) in
-                                       ( Meta t )}
+                                       ( Meta (Arg_qualifier_meta_tac t) )}
```
Could use `mk_meta_tac` here?

nikswamy replied:
Done

aseemr commented on src/typechecker/FStar.TypeChecker.Rel.fs:
```diff
@@ -312,6 +324,17 @@ let singleton wl prob smt_ok     = {wl with attempting=[prob]; smt_ok = smt_ok}
+let find_user_tac_for_uvar env (u:ctx_uvar) : option<sigelt> =
+    match u.ctx_uvar_meta with
+    | Some (Ctx_uvar_meta_attr a) ->
+      let hooks = Env.lookup_attr env Const.resolve_implicits_attr_string in
```
Do we need `Const.resolve_implicit_attrs_string` or can we directly lookup the attribute `a` in the `env`?

nikswamy replied:
Particularly if we intend to allow multiple attributes on a binder, I want a way to interpret some attributes as being relevant for resolve implicits and others remaining for other purposes. So, I see the `resolve_implicits a` as a marker indicating that `a` should be interpreted by the resolve implicits mechanism; other attributes could be handled in other ways.

aseemr commented on src/typechecker/FStar.TypeChecker.Rel.fs:
```diff
@@ -3860,22 +3933,137 @@ let resolve_implicits' env must_total forcelax g =
+            let _, tlhs, _ = env.type_of env_lax tp.lhs in
+            let goal_ty = U.mk_eq2 (env.universe_of env_lax tlhs) tlhs tp.lhs tp.rhs in
+            let goal, ctx_uvar, _ =
+              Env.new_implicit_var_aux reason tp.lhs.pos env goal_ty Allow_untyped None
```
Should this be `Allow_untyped` here? (Trying to understand why not `Strict`, could `Allow_untyped` be unsound.)

nikswamy replied:
I don't think it matters much, since this will be dispatched to a tactic anyway. And in the case of tactics, we do not re-typecheck solutions to goals. So, I left it as Allow_untyped for that reason, since in other usages, Strict means that we re-typecheck solutions.

nikswamy replied:
I made it Strict

aseemr commented on src/typechecker/FStar.TypeChecker.Rel.fs:
```diff
@@ -3860,22 +3933,137 @@ let resolve_implicits' env must_total forcelax g =
+          | TProb tp when tp.relation=EQ ->
+            let env = {env with gamma=tp.logical_guard_uvar.ctx_uvar_gamma} in
+            let env_lax = {env with lax=true; use_bv_sorts=true} in
+            let _, tlhs, _ = env.type_of env_lax tp.lhs in
```
Since we know that one of the sides in `EQ` is a flex, can't get the type from there (`ctx_uvar_typ`)?

nikswamy replied:
Maybe we should identify which side is flex and use it to compute the type.

I still would want to use env.type_of since even if it is flex, only its head term needs to be flex. It may be applied to a bunch of arguments. If it is indeed just an unapplied uvar, then type_of should just return the ctx_uvar_typ

nikswamy replied:
done.

nikswamy commented:
@aseemr , I'm not sure what this remark refers to:

> Can we also cleanup the previously added resolve_all_implicits_hook in this PR itself? The Rel support for this never made it to the master branch, but there are remnants of it in Env, Tactics, Pervasives etc. that we should get rid of.

Can you please clarify? thanks

aseemr commented:
Hi Nik, I was referring to the `env.try_solve_implicits_hook` which is disconnected in the current master branch (without `Rel` support). But I guess you are reusing it in this PR, so may be it's fine?

nikswamy commented:
FYI, I made the syntax

``(#[@@ resolve_implicits foobar])```

i.e., a single attribute with `@@`. The parser expects a `tmNoEq` so `[@@ f; g]` will not currently parse. That should make it future proof in case we want to add support for multiple attributes on a binder.

nikswamy commented:
I removed the sorting code, moved a lot of the logic of deferred implicits to tactics into DeferredImplicits.fs, so that the changes to Rel remain relatively small.

I moved SteelT.Effect and SteelT.Repro to tests/micro-benchmarks and have a tactic there that just checks that it gets the expected number of goals to solve (rather than checking any expected output).

Also added some comments etc.

nikswamy commented:
I also raise a warning on any use of an `#[@@]` binders, saying that this is experimental and may change.
